File: apps/radio-station/server/services/scheduler.ts
import type {ScheduledEntry, Track, Weights} from '../types.js';
import {WEEK_MS} from '../env.js';
import {createRng, type Rng} from './rng.js';
import {orbitPolicy, separationWindows, type OrbitPolicy, type SeparationWindows} from './clock.js';
import {groupTracks, numberOne, type TrackBuckets} from './buckets.js';
import {chooseBucket, pickFromBucket} from './pick.js';
import {shouldForceNumberOne} from './orbit.js';

export interface GenerateWeekInput {
    tracks: Track[];
    seed: string;
    startsAtMs: number;
    weights: Weights;
}

interface ClockState {
    buckets: TrackBuckets;
    windows: SeparationWindows;
    orbit: OrbitPolicy;
    top: Track | undefined;
    weights: Weights;
    weekStart: number;
}

function toEntry(track: Track, startsAtMs: number): ScheduledEntry {
    return {
        trackId: track.id,
        artist: track.artist,
        title: track.title,
        startsAtMs,
        durationMs: track.durationMs,
        rotation: track.rotation,
        era: track.era,
        rank: track.rank,
    };
}

function nextTrack(
    rng: Rng,
    clock: ClockState,
    now: number,
    lastByTrack: Map<string, number>,
    lastByArtist: Map<string, number>,
): Track {
    const {buckets, windows, orbit, top, weights, weekStart} = clock;
    if (shouldForceNumberOne(top, now, weekStart, lastByTrack, lastByArtist, windows, orbit, weights.goldLeak) && top) {
        return top;
    }
    const bucket = chooseBucket(rng, weights);
    return pickFromBucket(
        rng,
        buckets,
        bucket,
        now,
        lastByTrack,
        lastByArtist,
        windows,
        weights.temperature,
        orbit.minMs,
        top?.id,
    );
}

function buildClock(input: GenerateWeekInput): ClockState {
    const buckets = groupTracks(input.tracks);
    return {
        buckets,
        windows: separationWindows(input.weights.separation),
        orbit: orbitPolicy(input.weights),
        top: numberOne(buckets),
        weights: input.weights,
        weekStart: input.startsAtMs,
    };
}

export function generateWeek(input: GenerateWeekInput): ScheduledEntry[] {
    if (!input.tracks.length) throw new Error('catalog is empty');
    if (input.tracks.some((track) => track.durationMs <= 0)) throw new Error('catalog has a zero-duration track');
    const rng = createRng(input.seed);
    const clock = buildClock(input);
    const lastByTrack = new Map<string, number>();
    const lastByArtist = new Map<string, number>();
    const entries: ScheduledEntry[] = [];
    let now = input.startsAtMs;
    const end = input.startsAtMs + WEEK_MS;
    while (now < end) {
        const track = nextTrack(rng, clock, now, lastByTrack, lastByArtist);
        entries.push(toEntry(track, now));
        lastByTrack.set(track.id, now);
        lastByArtist.set(track.artist, now);
        now += track.durationMs;
    }
    return entries;
}

File: apps/radio-station/api/src/main/kotlin/radioapi/domain/Scheduler.kt
package radioapi.domain

fun generateWeek(tracks: List<Track>, seed: String, startsAtMs: Long, weights: Weights): List<ScheduledEntry> {
    if (tracks.isEmpty()) error("catalog is empty")
    if (tracks.any { it.durationMs <= 0 }) error("catalog has a zero-duration track")
    val rng = Rng(seed)
    val clock = buildClock(tracks, startsAtMs, weights)
    return fillWeek(rng, clock, startsAtMs)
}

private fun fillWeek(rng: Rng, clock: ClockState, startsAtMs: Long): List<ScheduledEntry> {
    val lastByTrack = mutableMapOf<String, Long>()
    val lastByArtist = mutableMapOf<String, Long>()
    val entries = mutableListOf<ScheduledEntry>()
    var now = startsAtMs
    val end = startsAtMs + WEEK_MS
    while (now < end) {
        val track = nextTrack(rng, clock, now, lastByTrack, lastByArtist)
        entries.add(toEntry(track, now))
        lastByTrack[track.id] = now
        lastByArtist[track.artist] = now
        now += track.durationMs
    }
    return entries
}

private fun nextTrack(
    rng: Rng,
    clock: ClockState,
    now: Long,
    lastByTrack: Map<String, Long>,
    lastByArtist: Map<String, Long>,
): Track {
    val forced = clock.top
    if (shouldForce(clock, now, lastByTrack, lastByArtist) && forced != null) return forced
    val bucket = chooseBucket(rng, clock.weights)
    return pickFromBucket(
        rng,
        clock.buckets,
        bucket,
        now,
        lastByTrack,
        lastByArtist,
        clock.windows,
        clock.weights.temperature,
        clock.orbit.minMs,
        forced?.id,
    )
}

private fun shouldForce(
    clock: ClockState,
    now: Long,
    lastByTrack: Map<String, Long>,
    lastByArtist: Map<String, Long>,
): Boolean = shouldForceNumberOne(
    clock.top,
    now,
    clock.weekStart,
    lastByTrack,
    lastByArtist,
    clock.windows,
    clock.orbit,
    clock.weights.goldLeak,
)

private fun toEntry(track: Track, startsAtMs: Long): ScheduledEntry = ScheduledEntry(
    trackId = track.id,
    artist = track.artist,
    title = track.title,
    startsAtMs = startsAtMs,
    durationMs = track.durationMs,
    rotation = track.rotation,
    era = track.era,
    rank = track.rank,
)

private data class ClockState(
    val buckets: TrackBuckets,
    val windows: SeparationWindows,
    val orbit: OrbitPolicy,
    val top: Track?,
    val weights: Weights,
    val weekStart: Long,
)

private fun buildClock(tracks: List<Track>, startsAtMs: Long, weights: Weights): ClockState {
    val buckets = groupTracks(tracks)
    return ClockState(
        buckets,
        separationWindows(weights.separation),
        orbitPolicy(weights),
        numberOne(buckets),
        weights,
        startsAtMs,
    )
}
